src/core/server/models/story/story.ts

```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
   // the null entries, the resulting array can not contain null.
   return results.filter((section) => section !== null).sort() as string[];
 }
+
+export async function markStoryForArchiving(
+  mongo: Db,
+  tenantID: string,
+  id: string,
+  now: Date
+) {
+  const result = await collection(mongo).findOneAndUpdate(
+    {
+      id,
+      tenantID,
+      $and: [
+        { $or: [{ isArchiving: { $exists: false } }, { isArchiving: false }] },
+        { $or: [{ isArchived: { $exists: false } }, { isArchived: false }] },
+      ],
+    },
+    {
+      $set: {
+        isArchiving: true,
+        closedAt: now,
+        updatedAt: now,
+        archivedAt: now,
+      },
+    },
+    {
+      returnOriginal: false,
+    }
+  );
+
+  return result.value;
+}
+
+export async function markStoryForUnarchiving(
+  mongo: Db,
+  tenantID: string,
+  id: string,
+  now: Date
+) {
+  const result = await collection(mongo).findOneAndUpdate(
+    {
+      id,
+      tenantID,
+      isArchiving: false,
+      isArchived: true,
+    },
+    {
+      $set: {
+        isArchiving: true,
+        closedAt: now,
+        updatedAt: now,
+      },
+    },
+    {
+      returnOriginal: false,
+    }
+  );
+
+  return result.value;
+}
+
+const getCommentCounts = (options: {
+  story: Readonly<Story>;
+  negate: boolean;
+}) => {
+  const {
+    story: { commentCounts },
+    negate,
+  } = options;
+  const multiplier = negate ? -1 : 1;
+
+  const result: RelatedCommentCounts = createEmptyRelatedCommentCounts();
+
+  if (commentCounts.action) {
+    for (const key in commentCounts.action) {
+      if (result.action.hasOwnProperty.call(commentCounts.action, key)) {
+        const value = commentCounts.action[key];
+        result.action[key] = value * multiplier;
+      }
```

Comment:
Best to invert this for more clarity:

```suggestion
      if (!Object.prototype.hasOwnProperty.call(commentCounts.action, key)) {
        continue;
      }
      
      const value = commentCounts.action[key];
      result.action[key] = value * multiplier;
```

Reply:
Apply the same changes below!

Reply:
👍